python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,261 @@
+def _(primitive_type, value: uuid.UUID) -> bytes:
+    return struct.pack(">QQ", (value.int >> 64) & 0xFFFFFFFFFFFFFFFF, value.int & 0xFFFFFFFFFFFFFFFF)
+
+
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(primitive_type, value: bytes) -> bytes:
+    return value
+
+
+@to_bytes.register(DecimalType)
+def _(primitive_type, value: Decimal) -> bytes:
+    value_as_tuple = value.as_tuple()
+    value_precision = len(value_as_tuple.digits)
+    value_scale = -value_as_tuple.exponent
+
+    if value_scale != primitive_type.scale:
+        raise ValueError(f"Cannot serialize value, scale of value does not match type {primitive_type}: {value_scale}")
+    elif value_precision != primitive_type.precision:
+        raise ValueError(f"Cannot serialize value, precision of value does not match type {primitive_type}: {value_precision}")
+
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        unscaled_value = int((value * 10**primitive_type.scale).to_integral_value(rounding=ROUND_HALF_DOWN))
+        min_num_bytes = ((unscaled_value).bit_length() + 7) // 8
+        return unscaled_value.to_bytes(min_num_bytes, "big", signed=True)
+
+
+@singledispatch
+def from_bytes(primitive_type, b: bytes):
+    """A generic function which converts bytes to a built-in python value
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        b(bytes): The bytes to convert
+    """
+    raise TypeError(f"Cannot deserialize bytes, type {primitive_type} not supported: {str(b)}")
+
+
+@from_bytes.register(BooleanType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<?", b)[0] != 0
+
+
+@from_bytes.register(IntegerType)
+@from_bytes.register(DateType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<i", b)[0]
+
+
+@from_bytes.register(LongType)
+@from_bytes.register(TimeType)
+@from_bytes.register(TimestampType)
+@from_bytes.register(TimestamptzType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<q", b)[0]
+
+
+@from_bytes.register(FloatType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<f", b)[0]
+
+
+@from_bytes.register(DoubleType)
+def _(primitive_type, b: bytes):
+    return struct.unpack("<d", b)[0]
+
+
+@from_bytes.register(StringType)
+def _(primitive_type, b: bytes):
+    return bytes(b).decode("utf-8")
+
+
+@from_bytes.register(UUIDType)
+def _(primitive_type, b: bytes):
+    return uuid.UUID(int=struct.unpack(">QQ", b)[0] << 64 | struct.unpack(">QQ", b)[1])
+
+
+@from_bytes.register(BinaryType)
+@from_bytes.register(FixedType)
+def _(primitive_type, b: bytes):
+    return b
+
+
+@from_bytes.register(DecimalType)
+def _(primitive_type, b: bytes):
+    integer_representation = int.from_bytes(b, "big", signed=True)
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        return Decimal(integer_representation) * Decimal(10) ** Decimal(-primitive_type.scale)
```

Comment:
```suggestion
        return Decimal(integer_representation) * (Decimal(10) ** Decimal(-primitive_type.scale))
```
Lets try to avoid relying on order of operations by python.